qemu/tests/driver_load.py

```diff
@@ -21,73 +22,94 @@ def run(test, params, env):
+        """
+        nic_index = len(vm.virtnet) - 1
+        session = vm.wait_for_login(nic_index=nic_index)
         output = session.cmd_output(cmd)
         driver_id = re.findall(pattern, output)
         if not driver_id:
-            raise error.TestFail("Didn't find driver info from guest %s"
-                                 % output)
+            raise exceptions.TestFail("Didn't find driver info from guest %s"
+                                      % output)
 
         driver_id = driver_id[0]
         if params["os_type"] == "windows":
             driver_id = '^&'.join(driver_id.split('&'))
+        session.close()
         return driver_id
 
-    error.context("Try to log into guest.", logging.info)
+    error_context.context("Try to log into guest.", logging.info)
     vm = env.get_vm(params["main_vm"])
     vm.verify_alive()
     timeout = float(params.get("login_timeout", 240))
-    # Use the last nic for send driver load/unload command
-    nic_index = len(vm.virtnet) - 1
-    session = vm.wait_for_login(nic_index=nic_index, timeout=timeout)
+    session = vm.wait_for_login(timeout=timeout)
 
-    driver_id_cmd = params["driver_id_cmd"]
     driver_id_pattern = params["driver_id_pattern"]
-    driver_load_cmd = params["driver_load_cmd"]
-    driver_unload_cmd = params["driver_unload_cmd"]
-
-    devcon = params.get("devcon")
-    if devcon:
-        error.context("Copy devcon.exe from winutils.iso to C:\\")
-        copy_devcon_cmd = params.get("devcon") % \
-            utils_misc.get_winutils_vol(session)
-        session.cmd(copy_devcon_cmd)
+    driver_id_cmd = utils_misc.set_winutils_letter(
+        session, params["driver_id_cmd"])
+    driver_load_cmd = utils_misc.set_winutils_letter(
+        session, params["driver_load_cmd"])
+    driver_unload_cmd = utils_misc.set_winutils_letter(
+        session, params["driver_unload_cmd"])
+    session.close()
 
     for repeat in range(0, int(params.get("repeats", 1))):
-        error.context("Unload and load the driver. Round %s" % repeat,
-                      logging.info)
-        error.context("Get driver info from guest", logging.info)
-        driver_id = check_driver(session, driver_id_cmd, driver_id_pattern)
-
-        error.context("Unload the driver", logging.info)
-        unload_driver(session, driver_unload_cmd, driver_id)
-        time.sleep(5)
-
-        error.context("Load the driver", logging.info)
-        load_driver(session, driver_load_cmd, driver_id)
-        time.sleep(5)
+        error_context.context("Unload and load the driver. Round %s" % repeat,
+                              logging.info)
+        logging.info("Get driver info from guest")
+        driver_id = get_driver_id(driver_id_cmd, driver_id_pattern)
+
+        try:
+            error_context.context("Unload the driver", logging.info)
+            unload_driver(driver_unload_cmd, driver_id)
+            time.sleep(5)
+        except exceptions.TestFail:
```

Comment:
I'm afraid driver is unloaded which will effect the following test.
snapshot should resolve this issue.
code have been udpated, could you review again